File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/services/BadgesService.kt
package me.nicolas.stravastats.domain.services


import me.nicolas.stravastats.domain.business.ActivityType
import me.nicolas.stravastats.domain.business.badges.*
import me.nicolas.stravastats.domain.services.activityproviders.IActivityProvider
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import tools.jackson.databind.DeserializationFeature
import tools.jackson.databind.json.JsonMapper
import tools.jackson.module.kotlin.KotlinModule
import java.nio.file.Files
import java.nio.file.Path

interface IBadgesService {
    fun getGeneralBadges(activityTypes: Set<ActivityType>, year: Int?): List<BadgeCheckResult>

    fun getFamousBadges(activityTypes: Set<ActivityType>, year: Int?): List<BadgeCheckResult>
}

@Service
internal class BadgesService(
    activityProvider: IActivityProvider,
) : IBadgesService, AbstractStravaService(activityProvider) {

    private val logger = LoggerFactory.getLogger(ActivityService::class.java)

    private val objectMapper = JsonMapper.builder()
        .addModule(KotlinModule.Builder().build())
        .disable(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES)
        .build()

    private val alpes: BadgeSet = loadBadgeSet("alpes", "famous-climb/alpes.json")

    private val pyrenees: BadgeSet = loadBadgeSet("pyrenees", "famous-climb/pyrenees.json")

    override fun getGeneralBadges(activityTypes: Set<ActivityType>, year: Int?): List<BadgeCheckResult> {
        logger.info("Checking general badges for $activityTypes in ${year ?: "all years"}")

        val activities = activityProvider.getActivitiesByActivityTypeAndYear(activityTypes, year)

        // TODO: handle case multiple activity types
        return when (activityTypes.firstOrNull()) {
            ActivityType.Ride -> {
                DistanceBadge.rideBadgeSet.check(activities) +
                        ElevationBadge.rideBadgeSet.check(activities) +
                        MovingTimeBadge.movingTimeBadgesSet.check(activities)
            }

            ActivityType.Hike -> {
                DistanceBadge.hikeBadgeSet.check(activities) +
                        ElevationBadge.hikeBadgeSet.check(activities) +
                        MovingTimeBadge.movingTimeBadgesSet.check(activities)
            }

            ActivityType.Run -> {
                DistanceBadge.runBadgeSet.check(activities) +
                        ElevationBadge.runBadgeSet.check(activities) +
                        MovingTimeBadge.movingTimeBadgesSet.check(activities)
            }

            else -> emptyList()
        }
    }

    override fun getFamousBadges(activityTypes: Set<ActivityType>, year: Int?): List<BadgeCheckResult> {
        logger.info("Checking famous badges for $activityTypes in ${year ?: "all years"}")

        val activities = activityProvider.getActivitiesByActivityTypeAndYear(activityTypes, year)

        // TODO: handle case multiple activity types
        return when (activityTypes.firstOrNull()) {
            ActivityType.Ride -> {
                alpes.check(activities) + pyrenees.check(activities)
            }

            else -> emptyList()
        }
    }

    private fun loadBadgeSet(name: String, climbsJsonFilePath: String): BadgeSet {
        val famousClimbs = loadFamousClimbs(climbsJsonFilePath)
        val famousClimbBadgeList = famousClimbs.flatMap { famousClimb ->
            famousClimb.alternatives.map { alternative ->
                FamousClimbBadge(
                    name = famousClimb.name,
                    label = "${famousClimb.name} from ${alternative.name}",
                    topOfTheAscent = famousClimb.topOfTheAscent,
                    start = famousClimb.geoCoordinate,
                    end = alternative.geoCoordinate,
                    difficulty = alternative.difficulty,
                    length = alternative.length,
                    totalAscent = alternative.totalAscent,
                    averageGradient = alternative.averageGradient
                )
            }
        }

        return BadgeSet(name, famousClimbBadgeList)
    }

    private fun loadFamousClimbs(climbsJsonFilePath: String): List<FamousClimb> {
        try {
            val filePath = resolveBadgeJsonFile(climbsJsonFilePath)
            if (filePath != null) {
                return objectMapper.readValue(filePath.toFile(), Array<FamousClimb>::class.java).toList()
            }

            val classpathResource = this::class.java.classLoader.getResourceAsStream(climbsJsonFilePath)
            if (classpathResource != null) {
                classpathResource.use { input ->
                    return objectMapper.readValue(input, Array<FamousClimb>::class.java).toList()
                }
            }

            logger.warn("Badge file not found for '$climbsJsonFilePath'. Badges set will be empty.")
            return emptyList()
        } catch (exception: Exception) {
            logger.error("Something was wrong while reading BadgeSet '$climbsJsonFilePath' : ${exception.message}")
            return emptyList()
        }
    }

    private fun resolveBadgeJsonFile(climbsJsonFilePath: String): Path? {
        val candidates = mutableListOf(
            Path.of(climbsJsonFilePath),
            Path.of("strava-cache").resolve(climbsJsonFilePath),
            Path.of("back-kotlin").resolve(climbsJsonFilePath),
        )

        val stravaCachePath = System.getenv("STRAVA_CACHE_PATH")
        if (!stravaCachePath.isNullOrBlank()) {
            candidates.add(Path.of(stravaCachePath).resolve(climbsJsonFilePath))
        }

        return candidates.firstOrNull { Files.isRegularFile(it) }
    }
}

File: back-go/internal/services/badgesService.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"

	"mystravastats/domain/badges"
	"mystravastats/domain/business"
)

var alpes = loadBadgeSet("alpes", "strava-cache/famous-climb/alpes.json")
var pyrenees = loadBadgeSet("pyrenees", "strava-cache/famous-climb/pyrenees.json")

// GetGeneralBadges returns the general badges for the given activity type and year
// The general badges are the ones that are not specific to a famous climb
func GetGeneralBadges(year *int, activityTypes ...business.ActivityType) []business.BadgeCheckResult {
	log.Printf("Checking general badges for %s in %v", activityTypes, year)

	activities := activityProvider.GetActivitiesByYearAndActivityTypes(year, activityTypes...)

	// TODO: handle case multiple activity types
	activityType := activityTypes[0]
	switch activityType {
	case business.Ride:
		return append(append(badges.DistanceRideBadgeSet.Check(activities),
			badges.ElevationRideBadgeSet.Check(activities)...),
			badges.MovingTimeBadgesSet.Check(activities)...)
	case business.Hike:
		return append(append(badges.DistanceHikeBadgeSet.Check(activities),
			badges.ElevationHikeBadgeSet.Check(activities)...),
			badges.MovingTimeBadgesSet.Check(activities)...)
	case business.Run:
		return append(append(badges.DistanceRunBadgeSet.Check(activities),
			badges.ElevationRunBadgeSet.Check(activities)...),
			badges.MovingTimeBadgesSet.Check(activities)...)
	default:
		return []business.BadgeCheckResult{}
	}
}

// GetFamousBadges returns the famous badges for the given activity type and year
// The famous badges are the ones that are specific to a famous climb
func GetFamousBadges(year *int, activityTypes ...business.ActivityType) []business.BadgeCheckResult {
	log.Printf("Checking famous badges for %s in %v", activityTypes, year)

	activities := activityProvider.GetActivitiesByYearAndActivityTypes(year, activityTypes...)

	activityType := activityTypes[0]

	switch activityType {
	case business.Ride:
		return append(alpes.Check(activities), pyrenees.Check(activities)...)

	default:
		return []business.BadgeCheckResult{}
	}
}

// loadBadgeSet loads the badge set from the given JSON file
func loadBadgeSet(name string, climbsJsonFilePath string) badges.BadgeSet {
	var famousClimbBadgeList []badges.Badge

	filePath, err := filepath.Abs(climbsJsonFilePath)
	if err != nil {
		log.Printf("Error getting absolute path: %v", err)
		return badges.BadgeSet{Name: name, Badges: famousClimbBadgeList}
	}

	file, err := os.Open(filePath)
	if err != nil {
		log.Printf("Error opening file: %v", err)
		return badges.BadgeSet{Name: name, Badges: famousClimbBadgeList}
	}
	defer func(file *os.File) {
		err := file.Close()
		if err != nil {
			log.Printf("Error closing file: %v", err)
		}
	}(file)

	byteValue, err := io.ReadAll(file)
	if err != nil {
		log.Printf("Error reading file: %v", err)
		return badges.BadgeSet{Name: name, Badges: famousClimbBadgeList}
	}

	var famousClimbs []badges.FamousClimb
	if err := json.Unmarshal(byteValue, &famousClimbs); err != nil {
		log.Printf("Error unmarshalling JSON: %v", err)
		return badges.BadgeSet{Name: name, Badges: famousClimbBadgeList}
	}

	for _, famousClimb := range famousClimbs {
		for _, alternative := range famousClimb.Alternatives {
			famousClimbBadgeList = append(famousClimbBadgeList, badges.FamousClimbBadge{
				Name:            famousClimb.Name,
				Label:           fmt.Sprintf("%s from %s", famousClimb.Name, alternative.Name),
				TopOfTheAscent:  famousClimb.TopOfTheAscent,
				Start:           famousClimb.GeoCoordinate,
				End:             alternative.GeoCoordinate,
				Difficulty:      alternative.Difficulty,
				Length:          alternative.Length,
				TotalAscent:     alternative.TotalAscent,
				AverageGradient: alternative.AverageGradient,
			})
		}
	}

	return badges.BadgeSet{Name: name, Badges: famousClimbBadgeList}
}
